### ISSUE TITLE
Contrast analyser doesn't factor in CSS shadow

### ISSUE THREAD

**Issue by DavidMacDonald** [dequelabs/axe-core#246]
A client who met contrast minimums by creating a 360 degree shadow around text. aXe measured foreground text and background but did not factor in the shadow... this would be tricky programming to remedy. It would require ensuring the shadow surrounds the text and that the shadow is sufficient, but I think it's worth investigating further, because I've had to over ride the aXe report.

**Comment by marcysutton** [comment]
We should be able to check for a CSS text-shadow if it's directly on the element (and not a pseudo-element). But how much blur, spread, etc. makes the shadow readable? We'd have to compare those values along with shadow color to the text and background colors to see if there is enough of a difference.

**Comment by Volker-E** [comment]
We've been running into this question ourselves without finding a distinct boundary.
Blur < 5px and offsets < 2px seemed to be approximations.

**Comment by dylanb** [comment]
@DavidMacDonald how did you determine that the contrast was sufficient using the shadow?

@Volker-E (side note: disabled elements do not have to meet the contrast minimums)

**Comment by DavidMacDonald** [comment]
That's where I use the Colour Contrast Analyser eyedroppers :)

**Comment by dylanb** [comment]
@DavidMacDonald What I mean by that question is, what relationship between the font-size/weight and the shadow-size leads to the same ability to read the text at the threshold contrast ratio? Do you have any research that backs up the claim that the shadow contrast is equivalent to the same background color?

**Comment by DavidMacDonald** [comment]
I have no research ... just the plain fact that it passes WCAG. There is no distinction between shadow and background. The text needs to have sufficient colour with its adjacent color (or outline) which we call background. When CSS shadow is used, it becomes the background as long as it is at 1px then it is background.

**Comment by dylanb** [comment]
@DavidMacDonald how do you determine whether the shadow definition actually results in a background that is at least 1px wide around the entire letter?

**Comment by WilcoFiers** [comment]
@DavidMacDonald I don't think it's quite as easy as you're saying it. Check out this codepen:
https://codepen.io/wilcofiers/pen/Kmxrra

In theory, all of these meet the color contrast, if you take the color of the text-shadow in place of the background. But offset and blur radius impact that quite a bit. Can you make a suggestion on how to deal with those?

**Comment by DavidMacDonald** [comment]
I agree, it would take some clever programming, and no one has done that before... some investigation of threshold blur etc... it would make you guys rock stars :)

This is not a simple bug fix, and may be a future investigation. But it is possible to test manually with eyedroppers, so perhaps it can be done automatically.

**Comment by dylanb** [comment]
@WilcoFiers @marcysutton WDYT - should we turn this into two different tickets:

1. look for a shadow and place it into review items
1. longer term, research looking for shadow width and doing this totally automatically

**Comment by WilcoFiers** [comment]
Nah. I think we can take this on in one go. I just have to figure out how to come up with a percantage of how much of the shadow's color we should mix with the background color. WCAG doesn't give us anything for that, but we can just put together a few examples and give those the numbers we feel are reasonable.

**Comment by DavidMacDonald** [comment]
the other trick is to ensure it's a 360 degree shadow, which requires two shadows, because it has a right left offset and a top bottom offset.

**Comment by tlobinger** [comment]
+1 on solving the issue

I think it should be easy/easier to calculate the merged colors if the assumption is the shadow is applied without offset. In that case the gradient can be calculated and what should matter is the color of the first pixel merged on top of the background. This new color becomes effectively an outline of the text..

So I would suggest for now to ignore shadow offsets and solving the issue for gradient backgrounds.
So limit the solution to solving this case https://jsfiddle.net/aq9Laaew/51325/

**Comment by dylanb** [comment]
@tlobinger I agree this would be a good first case to solve-for

@WilcoFiers comments?

**Comment by gulzarsingh591** [comment]
I am facing a similar issue while validating the text color(foreground) with text shadows(background) using axe.
In my case, we need to maintain the brand colors as well as the A11y contrast ratio. So to achieve both, we are referring a G18 technique which w3c is recommended. It says 

> If the background or the letters vary in relative luminance (or are patterned) then the background around the letters can be chosen or shaded so that the letters maintain a 4.5:1 contrast ratio with the background behind them even if they do not have that contrast ratio with the entire background.

Here is the link - https://www.w3.org/TR/WCAG20-TECHS/G18.html 

Not sure if your tool is programmed to pick the text shadows as background. Is the axe tool updated with this enhancement?

**Comment by WilcoFiers** [comment]
Here's another example of this:

Source: https://act-rules.github.io/rules/afw4f7#passed-example-4
```html
<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in a human language
</p>
```

**Comment by padmavemulapati** [comment]
Verified with the latest axe-core(3.5.5) code base , 
testfile - ` <p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in a human language
</p>` 
also verified on test file 
`<!DOCTYPE html>
<span id="span" style="text-shadow: 1px 1px 2px red,
    #f00 0 0 1em,
    0 0 0.2em rgba(0, 255, 0, 0.5);
">Hello world</span>
<script>
	window.addEventListener('load', function() {
		var span = document.getElementById('span')
		var style = window.getComputedStyle(span)
		console.log(style.getPropertyValue('text-shadow'))
		
	});
</script> `
![image](https://user-images.githubusercontent.com/47409579/88536753-fe7c0a80-d029-11ea-8e7b-9ba75463b6ab.png)

![image](https://user-images.githubusercontent.com/47409579/88536437-6ed65c00-d029-11ea-9a5c-6f82250c5d67.png)

**Comment by jeankaplansky** [comment]
included in internal release notes 7/27/2020
